### ISSUE TITLE
websocket support

### ISSUE THREAD

**Issue by cablehead** [imgix/levee#40]
Would be sweet, sweet, sweet

**Comment by thelinuxkid** [comment]
👍

**Comment by thelinuxkid** [comment]
# API Draft

## Handshake

A handshake is composed of a request from the client asking for an upgrade, a response from the server acknowledging the upgrade and a closing request or response (or both). Either side can start sending messages immediately after a successful handshake. The function names use the more popular term `upgrade`, even though the upgrade is just one part of the handshake.

ws.client_upgrade(conn, ...)
: Returns err

ws.server_upgrade(conn, ...)
: Returns err


## Encoding Messages

Message: data stream composed of one or more frames

Frame: payload plus control and extension data

FIN bit: indicates the final frame in a message

Opcodes:
  - 0x0: denotes a continuation frame
  - 0x1: denotes a text frame
  - 0x2: denotes a binary frame

Masking: clients must always mask data, servers must not.

### Unfragmented Message

ws.client_encode(buf, s, n)
: FIN bit set, opcode of 0x1 or 0x2, data masked

ws.server_encode(buf, s, n)
: FIN bit set, opcode of 0x1 or 0x2, data unmasked

### Fragmented Message

ws.client_frame(buf, s, n)
: FIN bit clear, opcode of 0x1 or 0x2, data masked

ws.client_frame_next(buf, s, n)
: FIN bit clear, opcode of 0x0, data masked

ws.client_frame_last(buf, s, n)
: FIN bit set, opcode of 0x0, data masked

ws.server_frame(buf, s, n)
: FIN bit clear, opcode of 0x1 or 0x2, data unmasked

ws.server_frame_next(buf, s, n)
: FIN bit clear, opcode of 0x0, data unmasked

ws.server_frame_last(buf, s, n)
: FIN bit set, opcode of 0x0, data unmasked


## Decoding Messages

TODO define frame object

ws.client_decode(stream)
: Returns err, frame

ws.server_decode(stream)
: Returns err, frame


## Control frames

These may be injected in the middle of a fragmented message but must not themselves be fragmented.

ws.close(buf)
: Connection close

ws.ping(buf)
: Ping

ws.pong(buf)
: Pong


# Alternative

ws.client(conn, ...)
: Returns err, client

ws.server(conn, ...)
: Returns err, server

`client` and `server` objects know if their frames need to be masked or not, eliminating the need for the prefixes `client_` and `server_` in the encoding/decoding functions.

**Comment by cablehead** [comment]
Should the upgrade hang off p.http?

**Comment by cablehead** [comment]
Just checking is `s, n` s|n` ?

**Comment by cablehead** [comment]
Maybe hold fire for the Alternative api for the moment? Just need a chance to play with it, to see how it looks..

**Comment by thelinuxkid** [comment]
Yes, it makes sense to hang the upgrade off `p.http`

`n` in `s, n` is there for completeness sake and to imply the length must be known ahead of time for each frame. In practice, `s:len()` would suffice. 

Here's a high-level overview of the framing
```
      0                   1                   2                   3
      0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
     +-+-+-+-+-------+-+-------------+-------------------------------+
     |F|R|R|R| opcode|M| Payload len |    Extended payload length    |
     |I|S|S|S|  (4)  |A|     (7)     |             (16/64)           |
     |N|V|V|V|       |S|             |   (if payload len==126/127)   |
     | |1|2|3|       |K|             |                               |
     +-+-+-+-+-------+-+-------------+ - - - - - - - - - - - - - - - +
     |     Extended payload length continued, if payload len == 127  |
     + - - - - - - - - - - - - - - - +-------------------------------+
     |                               |Masking-key, if MASK set to 1  |
     +-------------------------------+-------------------------------+
     | Masking-key (continued)       |          Payload Data         |
     +-------------------------------- - - - - - - - - - - - - - - - +
     :                     Payload Data continued ...                :
     + - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
     |                     Payload Data continued ...                |
     +---------------------------------------------------------------+
```

`Payload Data` is defined as the extension data concatenated with the application (user) data. For clients, the payload must be masked using a specific algorithm. The server uses the same algorithm to unmask the data. Specifically, the algorithm is:
```
   Octet i of the transformed data ("transformed-octet-i") is the XOR of
   octet i of the original data ("original-octet-i") with octet at index
   i modulo 4 of the masking key ("masking-key-octet-j"):

     j                   = i MOD 4
     transformed-octet-i = original-octet-i XOR masking-key-octet-j
```
I could apply this to byte(s) of the payload (as opposed to applying it to the entire payload at once), but, I have to test it. If so, we can use the `s|n` signature and send `n` bytes down the wire as we apply the algorithm. However, it might be easier to have the user chop-up the data and push the transformed data to `buf` inside each encoding function.

**Comment by thelinuxkid** [comment]
@cablehead do you think siphon's http parser should have a websocket counterpart?

**Comment by cablehead** [comment]
Yeah it definitely should, eventually. We can get to exposed API right for
it, with the parser in lua first though, and then replace to underlying
parser with a siphon one later.
